src/java/org/apache/cassandra/utils/concurrent/NonBlockingRateLimiter.java

```diff
@@ -0,0 +1,183 @@
+     *     in a "burst" prior to the limiter reaching a steady state. This allows the limiter 
+     *     to adhere more closely to the configured rate, as it can bring forward short
+     *     periods of inactivity if the rate of requests oscillates around the configured limit.
+     */
+    private final long burstNanos;
+
+    /** nanoseconds from start time corresponding to the next available permit */
+    private final AtomicLong nextAvailable = new AtomicLong();
+    
+    private volatile Ticker ticker;
+    
+    private volatile int permitsPerSecond;
+    
+    /** time in nanoseconds between permits on the timeline */
+    private volatile long intervalNanos;
+
+    public NonBlockingRateLimiter(int permitsPerSecond)
+    {
+        this(permitsPerSecond, DEFAULT_BURST_NANOS, Ticker.systemTicker());
+    }
+
+    @VisibleForTesting
+    public NonBlockingRateLimiter(int permitsPerSecond, long burstNanos, Ticker ticker)
+    {
+        this.startedNanos = ticker.read();
+        this.burstNanos = burstNanos;
+        setRate(permitsPerSecond, ticker);
+    }
+
+    public void setRate(int permitsPerSecond)
+    {
+        setRate(permitsPerSecond, Ticker.systemTicker());
+    }
+
+    @VisibleForTesting
+    public synchronized void setRate(int permitsPerSecond, Ticker ticker)
+    {
+        Preconditions.checkArgument(permitsPerSecond > 0, "permits/second must be positive");
+
+        this.ticker = ticker;
+        this.permitsPerSecond = permitsPerSecond;
+        intervalNanos = Math.max(TimeUnit.SECONDS.toNanos(1) / permitsPerSecond, 1);
+        nextAvailable.set(nanosElapsed());
+    }
+
+    /**
+     * @return the number of available permits per second
+     */
+    public int getRate()
+    {
+        return permitsPerSecond;
+    }
+
+    /**
+     * Reserves a single permit slot on the timeline which may not yet be available.
+     *
+     * @return microseconds until the reserved permit will be available (or zero if it already is)
+     */
+    public long reserveAndGetWaitMicros()
+    {
+        long nowNanos = nanosElapsed();
+
+        for (;;)
+        {
+            long prev = nextAvailable.get();
+            long interval = this.intervalNanos;
+
+            // Push the first available permit slot up to the burst window if necessary.
+            long firstAvailable = Math.max(prev, nowNanos - burstNanos);
+
+            // Advance the configured interval starting from the bounded previous permit slot.
+            if (nextAvailable.compareAndSet(prev, firstAvailable + interval))
+                // If the time now is before the first available slot, return the delay.  
+                return TimeUnit.NANOSECONDS.toMicros(Math.max(0,  firstAvailable - nowNanos));
+        }
+    }
+
+    /**
+     * Reserves a single permit slot on the timeline, but only if one is available.
+     *
+     * @return true if a permit is available, false if one is not
+     */
+    public boolean tryReserve()
```

Comment:
Having duplication between `reserveAndGetWaitMicros` and `tryReserve` makes it a little subtle to see that certain paths in the message handler are forcing permit reservation and processing queries and others not. While I don't love it, another option here would be unifying these two methods with something like the following:

```
/**
* Reserves a single permit slot on the timeline which may not yet be available.
*
* @return microseconds until the reserved permit will be available, 0 if available, -1 on failure
*/
public long tryReserve(boolean force)
{
    long nowNanos = nanosElapsed();

    for (;;)
    {
        long prev = nextAvailable.get();
        long interval = this.intervalNanos;

        // Push the first available permit slot up to the burst window if necessary.
        long firstAvailable = Math.max(prev, nowNanos - burstNanos);

        // If we haven't reached the time for the first available permit, we've failed to reserve.
        if (!force && nowNanos < firstAvailable)
            return -1;

        // Advance the configured interval starting from the bounded previous permit slot.
        // If another thread has already taken the next slot, retry.
        if (nextAvailable.compareAndSet(prev, firstAvailable + interval))
            return TimeUnit.NANOSECONDS.toMicros(Math.max(0,  firstAvailable - nowNanos));
    }
}
```

For the record: don't love that return API. Just wanted to surface that we should think twice before a) having two different places we keep our logic for "permit" reservation, and b) having a more clear API on "Reserve whether the time is right or not" vs. "Maybe Reserve" would be helpful.

Reply:
FWIW, I'm against this proposal. A boolean parameter is much poorer semantic distinction than good method naming. It's a pretty well established norm for `tryX` to mean _maybe_ do X (and return success/fail), and `x` to mean _always do X_ (and return the result), whereas the meaning of a boolean parameter has no strong compiler-enforced semantics and is easy to invert/confuse. Even within the method, in my opinion it is harder to parse the behaviour for any caller with the `force` parameter, though this is a weaker grounds for objection (and if we wanted to back both methods by this, I wouldn't complain if it were `private` - but I would still consider it inferior). 

TL;DR: We can bike shed the method names, but let's stick to enforcing semantics via method names where there's good optionality and no combinatorial explosion.

Reply:
I lean toward keeping things as they are here, partly for all the reasons Benedict mentions. We have two procedures that are implemented and documented in subtly different ways. Nominally, you might want to DRY this code out a bit to avoid having to make some future change in two places, but I would consider that almost a feature here. (i.e. I might *not* want to implicitly change both procedures with a code change to a piece of logic that is now common but might not be in the future. The guardrail for problems in either case seems like reasonable test coverage.)

Reply:
Completely fair, and the root of my "While I don't love it" disclaimer. More wanted to surface the discussion here as seeing the 2 permit allocation methods split on this last revision gave me brief pause, though it's clear why we need the split in flow.

I think having a `tryReserve` and a `reserve` vs. the current names would give more of the idiomatic symmetry I expected walking into the code here with the slight wrinkle of differing return types, but that's still quite clear as an API in my opinion.

Would also clarify the expectations on permit allocation a bit more in `PreV5Handlers.java` and `CQLMessageHandler.java` as in another comment here.

Reply:
I settled on `reserveAndGetDelay()` with a configurable delay unit. I can only go so terse ;)